src/server/twemcache-rs/src/worker/multi.rs

```diff
@@ -0,0 +1,358 @@
+                            self.wake_storage = true;
+                        }
+                        Err(ParseError::Incomplete) => {
+                            break;
+                        }
+                        Err(_) => {
+                            if session.deregister(&self.poll).is_err() {
+                                error!("Error deregistering");
+                            }
+                            session.close();
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn handle_new_sessions(&mut self) {
+        while let Ok(mut session) = self.session_receiver.try_recv() {
+            let pending = session.read_pending();
+            trace!("{} bytes pending in rx buffer for new session", pending);
+
+            // reserve vacant slab
+            let session_entry = self.sessions.vacant_entry();
+            let token = Token(session_entry.key());
+
+            // set client token to match slab
+            session.set_token(token);
+
+            // register tcp stream and insert into slab if successful
+            match session.register(&self.poll) {
+                Ok(_) => {
+                    session_entry.insert(session);
+                    if pending > 0 {
+                        // handle any pending data immediately
+                        if self.handle_data(token).is_err() {
+                            self.handle_error(token);
+                        }
+                    }
+                }
+                Err(_) => {
+                    error!("Error registering new socket");
+                }
+            };
+        }
+    }
+
+    pub fn message_sender(&self) -> Sender<Message> {
+        self.message_sender.clone()
+    }
+
+    pub fn session_sender(&self) -> Sender<Session> {
+        self.session_sender.clone()
+    }
+}
+
+impl EventLoop for MultiWorker {
+    fn get_mut_session(&mut self, token: Token) -> Option<&mut Session> {
+        self.sessions.get_mut(token.0)
+    }
+
+    fn handle_data(&mut self, token: Token) -> Result<(), ()> {
+        trace!("handling request for session: {}", token.0);
+        if let Some(session) = self.sessions.get_mut(token.0) {
+            loop {
+                if session.write_buffer.is_none() {
+                    break;
+                }
+                // TODO(bmartin): buffer should allow us to check remaining
+                // write capacity.
+                if session.write_pending() > MIN_BUFFER_SIZE {
+                    // if the write buffer is over-full, skip processing
+                    break;
+                }
+                // TODO(bmartin): using a trait here might make this nicer, eg:
+                // request.process(self, wbuf: &mut BytesMut, ... )
+                // need to check for performance impact.
+                match MemcacheParser::parse(&mut session.read_buffer) {
+                    Ok(request) => {
```

Comment:
This is shared with some of the code higher up, see if we can factor this out into a shared function